### DISCUSSION TITLE
Take 2: create model from JSON schema aka model_load_json_schema()

### DISCUSSION THREAD

**Discussion by fabiog1901** [pydantic/pydantic#6598] [Question] [answered]
Hello, with the V2 release I'd like to take the opportunity and dare to ask the eng team on a topic that was raised before for which only workarounds or very bespoke solutions where proposed.

In short, there is a method for dumping the Model to JSON, [model_json_schema](https://docs.pydantic.dev/latest/api/main/#pydantic.main.BaseModel.model_json_schema) but there is no counterpart, that is, a method that [re-]loads from JSON to a Model.

The use case is simple: I want to give my users the capability to store arbitrary artifacts. An artifact can be anything, any object: a car, a travel trip, a database cluster, a toilet, you name it. But it has to conform to a predefined schema for that object.

So an admin user would say, fine, we create the model schema (using the pydantic syntax). Then users create objects, making sure they conform with the model schema. This validation/sanitization involves loading the model from the saved json schema representation, and checking it against the user's input json payload.

Example model:

```python
class Car(BaseModel):
    id: int
    name: str | None = Field(max_length=50)
```

I'd save its JSON schema, below, in the database for long term storage. 

```json
{
    "properties": {
        "id": {
            "title": "Id",
            "type": "integer"
        },
        "name": {
            "anyOf": [
                {
                    "maxLength": 50,
                    "type": "string"
                },
                {
                    "type": "null"
                }
            ],
            "title": "Name"
        }
    },
    "required": [
        "id",
        "name"
    ],
    "title": "Car",
    "type": "object"
}
```

Then, when a user wants to create a car and passes a json payload (a dict) such as `{'id': 123, 'name': 'SuperCar', 'didntask': 'whatever'}`, I'd load the Pydantic Model dynamically using `create_model`, or a hypothetical `model_load_json_schema()`, feeding the above JSON schema, and validate/sanitize the user's input:

```python
# Car = create_model("Car", BaseModel, the_json_string)
Car = model_load_json_schema(the_json_string)

# validate/sanitize - i.e. conform to the schema
car = Car(**{'id': 123, 'name': 'SuperCar', 'didntask': 'whatever'})

print(car)
'''
id=123 name='SuperCar'
'''
```

Then I can safely store the `car.model_dump_json()` output to the database.

**Comment by mohandoz** [comment]
I'm actually working on this, i want the user to define his dynamic schema configuration from the UI and save them in DB to accept Dynamic api calls validated using the defined schema, still looking for a solution

**Comment by samuelcolvin** [comment, marked as answer]
Sure datamodel-code-generator is what you looking for https://docs.pydantic.dev/latest/integrations/datamodel_code_generator/.

Or you could write something that converted JSON Schema to pydantic core schema directly.

We won't be adding this functionality directly to pydantic.

**Reply by mohandoz** [reply]
Just stumbled on the projects minutes  ago Thanks!!

**Comment by h55nick** [comment]
@mohandoz @fabiog1901 I'm trying to do the same thing and datamodel_code_generator isn't going to be a solution for me.

This is where I'm at:
https://gist.github.com/h55nick/0e1d48b72ab9c5db33b89ba0443f41e7
but there are some issues with this current approach.

Would love to collaborate.


@samuelcolvin thanks for your work on Pydantic.

**Reply by drorata** [reply]
I'm also struggling with figuring out how to have the inverse of `model_json_schema`.  The use case in mind is to give the user some control on the data that will be processed and define a class during runtime based on a provided JSON/dict. The `runtime` part is what blocking the use of `datamodel-code-generator`. @h55nick would be happy to pick this up with you.

**Reply by drorata** [reply]
With a little bit of GenAI help I got the following code:


```python
from pydantic import BaseModel, Field, create_model

def create_model_from_json_schema(schema_name, schema_json):
  """
  Creates a Pydantic model from a given JSON Schema.

  Args:
    schema_name: The name of the model to be created.
    schema_json: The JSON Schema definition.

  Returns:
    The dynamically created Pydantic model class.
  """

  # Map JSON Schema types to Pydantic types
  type_mapping = {
      "string": str,
      "integer": int,
      "number": float,
      "boolean": bool, 
      # Add more mappings as needed (e.g., "array", "object", "null")
  }

  # Extract field definitions with type annotations
  field_definitions = {}
  for field_name, field_schema in schema_json["properties"].items():
    field_type = field_schema.get("type") 
    if field_type:
      python_type = type_mapping.get(field_type)
      if python_type:
        field_definitions[field_name] = (python_type, ...) 

  return create_model(schema_name, **field_definitions)

# Example usage
schema_json = {
    "type": "object",
    "properties": {
        "name": {"type": "string"},
        "age": {"type": "integer"},
    },
    "required": ["name"] 
}

User = create_model_from_json_schema("User", schema_json) 

# Validate data against the dynamically created model
user = User(**{"name": "Alice", "age": 30}) 
print(user)  # Output: name='Alice' age=30
```

It works except that I expected the output to be `User(name='Alice' age=30)`.

**Comment by spreeni** [comment]
Here a possible implementation for anyone stuck on this that covers nested relationships, list/dict item types, most common field constraints, descriptions and defaults.

```python
from typing import Any, Optional

from pydantic import BaseModel, Field, create_model

TYPE_MAPPING: dict[str, type] = {
    "string": str,
    "integer": int,
    "number": float,
    "boolean": bool,
    "object": dict,
    "array": list,
    "null": type(None),
}

CONSTRAINT_MAPPING: dict[str, str] = {
    "minimum": "ge",
    "maximum": "le",
    "exclusiveMinimum": "gt",
    "exclusiveMaximum": "lt",
    "inclusiveMinimum": "ge",
    "inclusiveMaximum": "le",
    "minItems": "min_length",
    "maxItems": "max_length",
}


def get_field_params_from_field_schema(field_schema: dict) -> dict:
    """Gets Pydantic field parameters from a JSON schema field."""
    field_params = {}
    for constraint, constraint_value in CONSTRAINT_MAPPING.items():
        if constraint in field_schema:
            field_params[constraint_value] = field_schema[constraint]
    if "description" in field_schema:
        field_params["description"] = field_schema["description"]
    if "default" in field_schema:
        field_params["default"] = field_schema["default"]
    return field_params


def create_model_from_schema(schema: dict) -> type[BaseModel]:  # noqa: C901
    """Create Pydantic model from a JSON schema generated by `Model.model_json_schema()`."""
    models: dict[str, type[BaseModel]] = {}

    def resolve_field_type(field_schema: dict) -> type[Any]:
        """Resolves field type, including optional types and nullability."""
        if "$ref" in field_schema:
            model_reference = field_schema["$ref"].split("/")[-1]
            return models.get(model_reference, Any)  # type: ignore[arg-type]

        if "anyOf" in field_schema:
            types = [TYPE_MAPPING.get(t["type"], Any) for t in field_schema["anyOf"] if t.get("type")]
            if type(None) in types:
                types.remove(type(None))
                if len(types) == 1:
                    return Optional[types[0]]
                return Optional[Union[tuple(types)]]  # type: ignore[return-value]
            else:
                return Union[tuple(types)]  # type: ignore[return-value]
        field_type = TYPE_MAPPING.get(field_schema.get("type"), Any)  # type: ignore[arg-type]

        # Handle arrays (lists)
        if field_schema.get("type") == "array":
            items = field_schema.get("items", {})
            item_type = resolve_field_type(items)
            return list[item_type]  # type: ignore[valid-type]

        # Handle objects (dicts with specified value types)
        if field_schema.get("type") == "object":
            additional_props = field_schema.get("additionalProperties")
            value_type = resolve_field_type(additional_props) if additional_props else Any
            return dict[str, value_type]  # type: ignore[valid-type]

        return field_type  # type: ignore[return-value]

    # First, create models for definitions
    definitions = schema.get("$defs", {})
    for model_name, model_schema in definitions.items():
        fields = {}
        for field_name, field_schema in model_schema.get("properties", {}).items():
            field_type = resolve_field_type(field_schema=field_schema)
            field_params = get_field_params_from_field_schema(field_schema=field_schema)
            fields[field_name] = (field_type, Field(**field_params))

        models[model_name] = create_model(model_name, **fields, __doc__=model_schema.get("description", ""))  # type: ignore[call-overload]

    # Now, create the main model, resolving references
    main_fields = {}
    for field_name, field_schema in schema.get("properties", {}).items():
        if "$ref" in field_schema:
            model_reference = field_schema["$ref"].split("/")[-1]
            field_type = models.get(model_reference, Any)  # type: ignore[arg-type]
        else:
            field_type = resolve_field_type(field_schema=field_schema)

        field_params = get_field_params_from_field_schema(field_schema=field_schema)
        main_fields[field_name] = (field_type, Field(**field_params))

    return create_model(schema.get("title", "MainModel"), **main_fields, __doc__=schema.get("description", ""))  # type: ignore[call-overload]
```

**Comment by pirate** [comment]
Really wish pydantic would reconsider adding this, Zod and ArkType both have it now: 

- https://zod.dev/json-schema?id=zfromjsonschema
- https://github.com/arktypeio/arktype/pull/1159

I'm trying to build [bidirectional event bridge](https://github.com/pirate/bbus) between Zod and Pydantic. It's all working from Pydantic -> Zod but the other directions is significantly more challenging without robust Pydantic support for making models from a JSONSchema at runtime.

I'd be happy to help contribute towards a PR if it helps @samuelcolvin. Pydantic is an amazing project and I think this feature could save a lot of people time. Even if not all of Pydantic's features would survive a JSONSchema roundtrip, having at least minimal support for basic types + nested objects/arrays/etc. would be great!
